codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -243,66 +246,38 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
+            if (openStream) {
+                stream.open(endOfStream);
+            }
 
-            if (lastDataWrite != null && !endOfStream) {
-                throw new IllegalStateException(
-                        "Sending non-trailing headers after data has been sent for stream: "
-                                + streamId);
+            // Pass headers to the flow-controller so it can maintain their sequence relative to DATA frames.
+            ChannelFuture future = flowController().sendFlowControlledData(ctx, stream,
+                    new FlowControlledHeaders(ctx, stream, headers, streamDependency, weight,
+                            exclusive, padding, endOfStream,  promise),
+                    promise);
+            future.addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    if (!future.isSuccess()) {
+                        // The write failed, handle the error.
+                        lifecycleManager.onException(ctx, future.cause());
+                    } else if (endOfStream) {
+                        lifecycleManager.closeLocalSide(stream, ctx.voidPromise());
```

Comment:
Definitely not right, fixed by making it listen on the passed in promise itself. That code is now moved into the FlowControlledPayload impl and is less verbose